### PR TITLE
Modify `singleQuote` option to allow single quotes in jsx

### LINKED ISSUE #1080: Provide way to use single quotes in jsx
I read [this comment](https://github.com/prettier/prettier/issues/73#issuecomment-271953751) saying that "I've never seen anyone write `<div className='a' />`". Well, here's someone 👋 

The same way `trailingComma` accepts a string, we could let `singleQuote` do as well. 
- `'all'` would use single quotes in js as well as jsx
- `'js'` (or `true`) would use single quotes in js but not in jsx
- `'jsx'` would use single quotes in jsx but not in js

### PR DESCRIPTION
**Pull request by trotzig**
We use jsx with single quotes in our application code:

  <Icon name='search' />

When I first tried out prettier, I thought that the `singleQuote` would
control jsx attributes as well. But I found out that it didn't.

It's likely that we are an outlier, so I didn't want to force
single-quotes all over. The solution I went with was to allow different
quote styles in regular js code and in jsx. The `singleQuote`
(--single-quote) option now takes a string that can have the following
values:

  "none" (default) - use double quotes everywhere
  "all" - use single quotes everywhere
  "js" - use single quotes in js code, double quotes in jsx
  "jsx" - use double quotes in jsx code, double quotes in js

Inspired by 5d13102, I kept support for the boolean option. It comes
with a deprecation warning, so people should find out that they have to
update soon.

Fixes #1080

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -259,7 +266,7 @@ Emacs users should see [this
 folder](https://github.com/jlongster/prettier/tree/master/editors/emacs)
 for on-demand formatting.
 
-### Vim 
+### Vim
```
Comment (trotzig) [inline README.md]: This is my editor cleaning out trailing whitespace on save.

File: bin/prettier.js [hunk 2]
```diff
@@ -122,7 +148,7 @@ const options = {
   printWidth: getIntOption("print-width"),
   tabWidth: getIntOption("tab-width"),
   bracketSpacing: argv["bracket-spacing"],
-  singleQuote: argv["single-quote"],
+  singleQuote: getSingleQuote(),
```
Comment (trotzig) [inline bin/prettier.js]: The whitespace was added by prettier (I ran `yarn run format` on this file).

File: bin/prettier.js [hunk 3]
```diff
@@ -218,7 +245,7 @@ if (stdin) {
     let input;
     try {
       input = fs.readFileSync(filename, "utf8");
-    } catch(e) {
+    } catch (e) {
```
Comment (trotzig) [inline bin/prettier.js]: The whitespace was added by prettier (I ran `yarn run format` on this file).

File: docs/index.html [hunk 4]
```diff
@@ -88,7 +88,7 @@
     <label class="title"><a href="https://github.com/jlongster/prettier">prettier v<span class="version"></span></a></label>
     <label><input type="number" value="80" id="printWidth"></input> printWidth</label>
     <label><input type="number" value="2" id="tabWidth"></input> tabWidth</label>
-    <label><input type="checkbox" id="singleQuote"></input> singleQuote</label>
+    <label>singleQuote <select id="singleQuote"><option value="none">none</option><option value="js">js</option><option value="jsx">jsx</option><option value="all">all</option></select></label>
```
Comment (trotzig) [inline docs/index.html]: I haven't found a way to test this change.

Comment (vjeux) [inline docs/index.html]: Open `docs/index.html`

Comment (vjeux) [inline docs/index.html]: And

```js
yarn build:docs
```

to regenerate the js

Comment (trotzig) [inline docs/index.html]: Ah, thanks. I'm so used to things having to be transpiled/compiled these days so I didn't even consider opening the html file directly an option...

Comment (trotzig) [inline docs/index.html]: I had a look at git history for `docs/prettier.min.js`, and it looks like you've been regenerating it for each release. I'm leaving it as-is in this PR, and if you think it's a good idea, I can add something to the contributing section in the readme about the workflow for the docs.

**Comment by trotzig** [comment]
I pushed a commit to update the broken yarn.lock file. And then I had to add back some whitespace that my git client stripped out automatically in order to get the jsx snapshot tests passing. 

This should be ready to go if you think it's in a good shape.

**Comment by jlongster** [comment]
Let's continue the discussion in #1080. I really, really think we shouldn't add this config option.

**Comment by rattrayalex** [comment]
Related issue closed, so closing this too. 

The implementation looked great, though, and very thorough; @trotzig would love to see more PR's in the future!

**Comment by jlongster** [comment]
See https://github.com/prettier/prettier/issues/1080#issuecomment-289599366. Thank for the discussing this @trotzig and sorry I decided against it... Also, I'm only one person, so if @vjeux or any other maintainers disagree we can reopen this.

**Pull request closed without merging**